[LV] Add a statistic for early exit vectorization

annamthomas opened the pull request:
We currently do not vectorize the epilog loops with early-exits, but the
stats are updated there as well for completeness.

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -10291,6 +10293,9 @@ bool LoopVectorizePass::processLoop(Loop *L) {
         auto ExpandedSCEVs = LVP.executePlan(EPI.MainLoopVF, EPI.MainLoopUF,
                                              *BestMainPlan, MainILV, DT, false);
         ++LoopsVectorized;
+        // TODO: Currently unsupported for early exits.
+        if (BestPlan.hasEarlyExit())
+          ++LoopsEarlyExitVectorized;
```
Can we just increment the statistic in a single place, like `executePlan`?

fhahn reviewed (commented):
Might also be worth adding a test for the new statistic

fhahn commented on llvm/lib/Transforms/Vectorize/LoopVectorize.cpp:
```diff
@@ -7324,6 +7325,11 @@ DenseMap<const SCEV *, Value *> LoopVectorizationPlanner::executePlan(
          "Trying to execute plan with unsupported VF");
   assert(BestVPlan.hasUF(BestUF) &&
          "Trying to execute plan with unsupported UF");
+  ++LoopsVectorized;
```
Would be good to leave the other increments as-is for now. 

Doesn't this change now increments LoopsVectorized twice per loop if the epilogue gets vectorized?

annamthomas committed:
[LV] Add a statistic for early exit vectorization

Add statistic LoopsEarlyExitVectorized

fhahn commented on llvm/test/Transforms/LoopVectorize/vect.stats.ll:
```diff
@@ -1,12 +1,12 @@
-; RUN: opt < %s -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -debug-only=loop-vectorize --disable-output -stats -S 2>&1 | FileCheck %s
+; RUN: opt < %s -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -debug-only=loop-vectorize -enable-early-exit-vectorization=1 --disable-output -stats -S 2>&1 | FileCheck %s
```
`=1` shouldn't be needed because it's the default now, but I guess it also doesn't hurt.

fhahn commented on llvm/test/Transforms/LoopVectorize/vect.stats.ll:
```diff
@@ -31,6 +31,39 @@ for.end:                                          ; preds = %entry, %for.body
+  %ind.next = add i8 %ind, 1
+  %conv = zext i8 %ind.next to i32
+  %gep.ind.next = add i64 %gep.ind, 1
+  %cmp = icmp ult i32 %conv, %end.clamped
```
might as well use a simpler case, with constant here, i64 compare to remove the unnecessary zext.

Also can use a single IV?

fhahn reviewed (approved):
LGTM, left a few suggestions to simplify the test case, as it seems like a slightly simpler one would suffice.

fhahn reviewed (changes requested):
Looks like there are some test failures, possibly for cases when there are no exit blocks (perhaps for interleave groups which require the scalar epilogue?)

Suspect we may need to check if there's at least one exit block in `hasEarlyExit`

annamthomas committed:
[VPlan] Update hasEarlyExit check to consider no-exit block case

If we call this API during vectorization without any exit blocks, we
need to first check there is atleast one exit block.

fhahn commented on llvm/lib/Transforms/Vectorize/VPlan.h:
```diff
@@ -4247,7 +4247,8 @@ class VPlan {
   bool hasEarlyExit() const {
-    return ExitBlocks.size() > 1 || ExitBlocks[0]->getNumPredecessors() > 1;
+    return ExitBlocks.size() > 1 ||
+           (ExitBlocks.size() > 0 && ExitBlocks[0]->getNumPredecessors() > 1);
```
```suggestion
           (ExitBlocks.size() == 1 && ExitBlocks[0]->getNumPredecessors() > 1);
```

In that case we must have a single exit block, right?

david-arm commented on llvm/test/Transforms/LoopVectorize/vect.stats.ll:
```diff
@@ -1,12 +1,12 @@
-; RUN: opt < %s -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -debug-only=loop-vectorize --disable-output -stats -S 2>&1 | FileCheck %s
+; RUN: opt < %s -passes=loop-vectorize -force-vector-interleave=4 -force-vector-width=4 -debug-only=loop-vectorize -enable-early-exit-vectorization --disable-output -stats -S 2>&1 | FileCheck %s
```
The flag `enable-early-exit-vectorization` can be removed now, since it's on by default.

david-arm reviewed (approved):
LGTM with comment addressed!

fhahn reviewed (approved):
LGTM, thanks

annamthomas committed:
Merge branch 'main' into EE-Stat
